File: leetcode/91.decode-ways.py
# @leet start
class Solution:
    def numDecodings(self, s: str) -> int:
        n = len(s)
        last = int(s[n - 1] != "0")
        second_last = 1

        for i in range(n - 2, -1, -1):
            if s[i] == "0":
                curr = 0
            else:
                curr = last

                if s[i] == "1" or s[i] == "2" and s[i + 1] <= "6":
                    curr += second_last

            second_last = last
            last = curr

        return last


# @leet end

File: leetcode/91.decode-ways.js
// @leet start
/**
 * @param {string} s
 * @return {number}
 */
var numDecodings = function (s) {
  let prev = 1;
  let prevPrev = 0;

  for (let i = s.length - 1; i >= 0; i--) {
    let curr = 0;

    if (s[i] !== "0") {
      curr = prev;

      if (
        i < s.length - 1 &&
        (s[i] === "1" || (s[i] === "2" && s[i + 1] < "7"))
      ) {
        curr += prevPrev;
      }
    }

    [prev, prevPrev] = [curr, prev];
  }

  return prev;
};
// @leet end
